### PR TITLE
Haim/issues fix

### PR DESCRIPTION
**Pull request by haimkastner**
1) Cache mechanism #20 .
2) Logger.

### CODE REVIEW

File: src/core/cache.ts [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+  /**
+   * Get value by key.
+   * @param key Key to get for.
+   * @returns The value, or 'undefined' if not exist.
+   */
+  public async get(key: string | number): Promise<any> {
+    return this._nodeCache.get(key);
+  }
+
+  /** Save or set value map by key to cache. */
```
Comment (eranshmil) [inline src/core/cache.ts]: JSDoc? :)

File: src/security/authentication.ts [hunk 2]
```diff
@@ -1,13 +1,10 @@
...
 
 import { AuthenticatedRequest } from '../models';
 import { checkReporterKey } from '../data';
+import { Cache, logger } from '../core';
 
-const reportersCache = new NodeCache({
-  stdTTL: 60 * 60 * 2, // Every 2 hours reread key from DB.
-  checkperiod: 60 * 30 // Clear old cache every 30 minutes.
-});
+const reportersAuthCache = new Cache(2 * 60 * 60);
```
Comment (eranshmil) [inline src/security/authentication.ts]: Cache for this one isn't necessary, the report route wouldn't be called so frequently.

File: src/models/user-status.model.ts [hunk 3]
```diff
@@ -6,30 +6,29 @@ import { Platform, Status, Reason } from '../core';
 
 @Entity({ name: 'user_statuses' })
 export class UserStatus extends AuthenticatedRequest {
-  @PrimaryGeneratedColumn()
-  private id: number;
+  @PrimaryGeneratedColumn() private id: number;
 
-  @Column({ name: 'user_id', type: 'varchar', length: 30 })
+  @Column({ name: 'user_id', type: 'varchar', length: 30, nullable: false })
```
Comment (eranshmil) [inline src/models/user-status.model.ts:15]: You can add the `postId` also in this PR.

File: src/controllers/user-statuses-controller.ts [hunk 4]
```diff
@@ -21,47 +21,29 @@ export class UserStatusesController extends Controller {
...
-    //   if (!(userId in bots)) {
-    //     botsCache.set(`${platform}:${userId}`, 'NOT_EXIST');
-    //   }
-    // }
-
-    return userStatuses;
+    /** Try to retrieve users status from the cache before reading data from the DB. */
+    const usersStatusMap: UserStatusMap = {};
+    for (const userId of userIds) {
+      const cacheResult = await usersCache.get(`${platform}:${userId}`);
```
Comment (eranshmil) [inline src/controllers/user-statuses-controller.ts:27]: For the next version, we should change that logic to use `mget` and some other structure for the values.
What do you think?

**Review by eranshmil — CHANGES_REQUESTED** [review]

File: src/core/cache.ts (hunk 1 above)
Comment (haimkastner) [reply src/core/cache.ts]: Yes... I fixed it in 8ae4d3da051b02ae74eaf205572a24405a8b6cef

File: src/security/authentication.ts (hunk 2 above)
Comment (haimkastner) [reply src/security/authentication.ts]: I removed it in 8ae4d3da051b02ae74eaf205572a24405a8b6cef, but it was fun.

File: src/models/user-status.model.ts (hunk 3 above)
Comment (haimkastner) [reply src/models/user-status.model.ts:15]: Implemented with 581bf277cfcd6fa39ab547044d2b00ccc8a109f1

**Review by eranshmil — APPROVED** [review]

**Pull request merged**